src/org/rascalmpl/library/lang/paths/Windows.rsc

```diff
@@ -0,0 +1,113 @@
+resolves to the same file on system A via any ((module::IO)) function. 
+* and nothing more. No normalization, no interpretatioon of `.` and `..`, no changing of cases. 
+This is left to downstream processors of `loc` values, if necessary. The current transformation
+is purely syntactical, and tries to preserve the semantics of the path as much as possible.
+}
+module lang::paths::Windows
+
+import IO;
+import util::SystemAPI;
+
+lexical WindowsPath
+    = unc              : Slash Slash Slashes? PathChar* hostName Slashes PathChar* shareName Slashes WindowsFilePath path
+    | absolute         : Drive drive ":" Slashes WindowsFilePath path
+    | driveRelative    : Drive drive ":" WindowsFilePath path
+    | directoryRelative: Slash WindowsFilePath
+    | relative         : WindowsFilePath path         
+    ;
+
+lexical PathChar = !([\a00-\a20\< \> : \" | ? * \\ /] - [\ ]);
+
+lexical PathSegment
+    = current: "."
+    | parent : ".."
+    | name   : PathChar+ \ ".." \ "."
+    ;
+
+lexical Drive = [A-Za-z];
+
+lexical Slashes = Slash+ !>> [\\/];
+
+lexical Slash = [\\/];
+
+lexical WindowsFilePath = {PathSegment Slashes}* segments Slashes? [\ .] !<< (); // only the last segment must not end in spaces.
+
+import ParseTree;
+
+@synopsis{Convert a windows path literal to a source location URI}
+@description{
+1. parses the path using the grammar for ((WindowsPath))
+2. takes the _literal_ name components using string interpolation `"<segment>"`. This means no decoding/encoding happens at all while extracting
+hostname, share name and path segment names. Also all superfluous path separators are skipped.
+3. uses `loc + str` path concatenation with its builtin character encoding to construct the URI. Also
+the right path separators are introduced. 
+}
+loc parseWindowsPath(str input, loc src=|unknown:///|) = mapPathToLoc(parse(#WindowsPath, input, src));
+
+@synopsis{UNC}
+loc mapPathToLoc((WindowsPath) `<Slash _><Slash _><Slashes? _><PathChar* hostName><Slashes _><PathChar* shareName><Slashes _><WindowsFilePath path>`)
+    = (|unc://<hostName>/| + "<shareName>" | it + "<segment>" | segment <- path.segments );
+
+@synopsis{Absolute: given the drive and relative to its root.}
+loc mapPathToLoc((WindowsPath) `<Drive drive>:<Slashes _><WindowsFilePath path>`) 
+    = (|file:///<drive>:/| | it + "<segment>" | segment <- path.segments);
+
+@synopsis{Drive relative: relative to the current working directory on the given drive.}
+loc mapPathToLoc((WindowsPath) `<Drive drive>:<WindowsFilePath path>`) 
+    = (|file:///<drive>:.| | it + "<segment>" | segment <- path.segments);
+
+@synopsis{Directory relative: relative to the root of the current drive.}
+loc mapPathToLoc((WindowsPath) `<Slash _><WindowsFilePath path>`) 
+    = (|cwdrive:///| | it + "<segment>" | segment <- path.segments);
+
+@synopsis{Relative to the current working directory on the current drive.}
+loc mapPathToLoc((WindowsPath) `<WindowsFilePath path>`) 
```

Comment:
all these 4 functions are the same, with the exception of the root uri, can we refactor that? so that it becomes something like:

```
loc mapPathToLoc((WindowsPath) `<Drive drive>:<WindowsFilePath path>`) = build(|file:///<drive>:.|, path);
loc mapPathToLoc((WindowsPath) `<WindowsFilePath path>`) = build(|cwd:///|, path);
...
```
?

Reply:
Definitely possible and would make the code less wide. I'll have a look.